libs/src/api/Users.ts

```diff
@@ -1002,25 +1002,34 @@ export class Users extends Base {
     const asyncFn = async () => {
       while (true) {
         const encodedUserId = Utils.encodeHashId(userId)
-        const replicaSet = await this.discoveryProvider.getUserReplicaSet({
-          encodedUserId: encodedUserId!,
-          blockNumber
-        })
-        if (
-          replicaSet &&
-          replicaSet.primarySpID === replicaSetSPIDs[0] &&
-          replicaSet.secondary1SpID === replicaSetSPIDs[1] &&
-          replicaSet.secondary2SpID === replicaSetSPIDs[2]
-        ) {
-          break
+        try {
```

Comment:
this code is still hard to follow
1. we should add comment that the `getUserReplicaSet()` call will throw an error if `blockNumber` hasn't been indexed
2. this logic still seems partially incorrect to me - what happens `if (!replicaSet)` -> even tho blocknumber was indexed, we'll still keep retrying

also can you add a function comment indicating the different possible outcomes? as i understand them:
- replicaset matches at blocknum -> return null
- no replicaset at blocknum -> throw error
- mismatched replicaset at block -> throw error
- timeout exceeded before replica set indexed -> throw error